CLDSRV-89 - update kubernetes readiness/liveness probe route

lavigne958 opened the pull request:
Update the readiness and liveness route for kubernetes.
The operator will configure the pod to check for readiness
and liveness on the same port as for the metrics.

In this commit allow the new routes on the metric code path.

Commit:
ZENKO-3725 - update kubernetes readiness/liveness probe route

Update the readiness and liveness route for kubernetes.
The operator will configure the pod to check for readiness
and liveness on the same port as for the metrics.

In this commit allow the new routes on the metric code path.

bert-e commented:
# Hello lavigne958,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/cloudserver/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-89 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.6`


Please check the `Fix Version/s` of CLDSRV-89, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jbertran commented on lib/utilities/monitoringHandler.js:
```diff
@@ -190,7 +190,9 @@ function monitoringHandler(clientIP, req, res, log) {
     }
     const monitoring = (req.url === '/metrics');
-    if (!monitoring) {
+    const healthcheck = (req.url === '/healthcheck');
```
Minor: this is a switch with fallthrough on every case, and `MethodNotAllowed` as default.

lavigne958 replied:
don't need anymore

francoisferrand commented on lib/utilities/monitoringHandler.js:
```diff
@@ -190,7 +190,9 @@ function monitoringHandler(clientIP, req, res, log) {
+    if (!monitoring && !healthcheck && !healthcheckdeep) {
         return monitoringEndHandler(errors.MethodNotAllowed, []);
     }
     return routeHandler(req, res, monitoringEndHandler);
```
I don't think this works.
`monitoringHandler()` function is not expected to route traffic : it actually just handles the /metrics path, and will always return the metrics.

(granted, maybe this `routeHandler()` function is not clearly named; but it is a local/private function, and just means "handle this route")

Healthchecks are handled by `utilities/healthcheckHandler`, and should not go through monitoring : this "dispatch" may be done in `routeAdminRequest()` in `server.js`.

lavigne958 replied:
will do,.

lavigne958 replied:
done.

francoisferrand reviewed (changes requested):
Please also update documentation (`Healthchecks.md`), existing tests (at least metrics test rely on this) and add specific tests for this feature, if not already present.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

Commit:
CLDSRV-89 - update liveness/readiness path

Use the shorter path `/healthcheck`, use the dedicated function `healthckeckhandler`
with it.

Commit:
CLDSRV-89 - update doc, tests to new health check route

Add branch `user/*` to trigger a CI run, in order for everyone
to be able to run CI on personal branches.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-89 contains:


* `8.3.6`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.7`


Please check the `Fix Version/s` of CLDSRV-89, or the target
branch of this pull request.

lavigne958 commented:
read to review.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

francoisferrand commented on Healthchecks.md:
```diff
@@ -1,6 +1,6 @@
 # S3 Healthcheck
 
-Scality S3 exposes a healthcheck route `/_/healthcheck` which returns a
+Scality S3 exposes a healthcheck route `/healthcheck` which returns a
```
should indicate this is not on the same port as s3, actually.

lavigne958 replied:
actually not, in the ticket it is defined that we put it on a different port, it should be on the metrics port, but sure we should indicate the port to use.

francoisferrand replied:
yes exactly my point : looking at it here we may think that healthcheck is on the same port as S3, which is not the case. So we should indicate it is on this metrics port (or admin, not sure how it is called in the config).

lavigne958 replied:
will do.

lavigne958 replied:
done.

francoisferrand commented on eve/main.yml:
```diff
@@ -2,7 +2,7 @@
 
 branches:
-  feature/*, documentation/*, improvement/*, bugfix/*, w/*, q/*, hotfix/*, dependabot/*:
+  feature/*, documentation/*, improvement/*, bugfix/*, w/*, q/*, hotfix/*, dependabot/*, user/*:
```
should this be commited?

lavigne958 replied:
yep, so everyone can create a `user/` branch to test things on the CI and not have to make `feature/` or `documentation/` branches.

francoisferrand commented on lib/management/push.js:
```diff
@@ -277,7 +277,7 @@ function addOverlayMessageListener(callback) {
 function startPushConnectionHealthCheckServer(cb) {
     const server = http.createServer((req, res) => {
-        if (req.url !== '/_/healthcheck') {
+        if (req.url !== '/healthcheck') {
```
not sure this is related.
I don't know what this "push connection" is, but it seems to be a dedicated server : so it is not needed for our purpose I think, and changing it here would break the other end of the push channel?

lavigne958 replied:
it seems to be a server that has listent to a websocket and pushes some metrics somewhere. it has the healthcheck route and I assumed that route is checked. if it is not, we could leave it here untouched.

francoisferrand replied:
i am not sure, but seeing the same value is not proof enough : since this `/_/healthcheck` is used everywhere...
hopefully someone knowledgeable on the subject will be able to confirm either way....

lavigne958 replied:
right, wait for confirmation from someone else then regarding the outcome I will push a new commit.
I'll try to limit the number of commits as it is very hard to make them land and have a green build

lavigne958 replied:
After looking around, this piece of code is not called by cloudserver. it is a dedicated server that can be run manually (manually => yarn run start_metrics_server) may be it is used by other components or simply unused at all.

lavigne958 replied:
I think I should update the route to makes sure it matches the same route name as expected by zenko in general

francoisferrand replied:
i confirmed with Rached that this is not used in Artesca (and should generally be removed, it was related to Orbit and the "cloud management" feature or Zenko 1)

lavigne958 replied:
I am not too sure, this is run in artesca, or at least when I run a local cluster I have a pod running, I will try on a fresh artesca, but I think it runs. we could update it so it uses the `/live` route even though it will be removed for now it is here and should follow the update ?

Commit:
CLDSRV-89 - rename route `healthcheck` to `live`

Rename the route `/healthcheck` to `/live` to match
kubernetes probe name *liveness*.

Update the route `/healthcheck/deep` to `/ready` to match
kubernetes probe name *readiness*

lavigne958 commented:
Updated the route name to match new names:
- liveness probe: `/live`
- readiness probe: `/ready`

Commit:
CLDSRV-89 - specify health check port in documentation

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
CLDSRV-89 - revert push server health check route

Revert the route used to check the health of the push server.
This route is not being updated in zenko-operator, so it should
not change to make sure it matches what the operator deploys

lavigne958 commented:
revert the changes on the push server.
Waitting for green CI

lavigne958 commented:
Needed to rebase my branch to include the fix to pass the CI.

lavigne958 commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.3`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`
* `stabilization/7.4.10`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.3`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`
* `stabilization/7.4.10`



Please check the status of the associated issue CLDSRV-89.

Goodbye lavigne958.
